sdks/java/io/cdap/src/main/java/org.apache.beam.sdk.io.cdap/CdapPlugin.java

```diff
@@ -0,0 +1,173 @@
+
+  private enum Format {
+    INPUT("mapreduce.job.inputformat.class", "key.class", "value.class"),
+    OUTPUT("mapreduce.job.outputformat.class", "key.class", "value.class");
+
+    private final String formatClass;
+    private final String keyClass;
+    private final String valueClass;
+
+    Format(String formatClass, String keyClass, String valueClass) {
+      this.formatClass = formatClass;
+      this.keyClass = keyClass;
+      this.valueClass = valueClass;
+    }
+  }
+
+  private Boolean isUnbounded;
+  private Format format;
+
+  private final Class<?> cdapPluginClass;
+
+  private Configuration hadoopConf;
+  private BatchContextImpl context;
+
+  protected Class<?> formatClass;
+  protected Class<?> keyClass;
+  protected Class<?> valueClass;
+  protected Class<?> formatProviderClass;
+
+  private T cdapPluginObj;
+
+  public CdapPlugin(Class<T> cdapPluginClass) {
+    this.cdapPluginClass = cdapPluginClass;
+
+    // Determine is unbounded or bounded
+    for (Annotation annotation : cdapPluginClass.getDeclaredAnnotations()) {
+      if (annotation.annotationType().equals(Plugin.class)) {
+        String pluginType = ((Plugin) annotation).type();
+        isUnbounded = pluginType != null && pluginType.startsWith("streaming");
+      }
+    }
+    if (isUnbounded == null) {
+      throw new IllegalArgumentException("CDAP plugin class must have Plugin annotation!");
+    }
+
+    // Init context and determine input or output
+    Class<?> contextClass = null;
+    for (Method method : cdapPluginClass.getDeclaredMethods()) {
+      if (method.getName().equals("prepareRun")) {
+        contextClass = method.getParameterTypes()[0];
+      }
+    }
+    if (contextClass == null) {
+      throw new IllegalStateException("Cannot determine context class");
+    }
+
+    if (contextClass.equals(BatchSourceContext.class)) {
+      context = new BatchSourceContextImpl();
+      format = Format.INPUT;
+    } else if (contextClass.equals(BatchSinkContext.class)) {
+      context = new BatchSinkContextImpl();
+      format = Format.OUTPUT;
+    } else {
+      // TODO: context = new StreamingSourceContextImpl();
+      format = Format.INPUT;
+      context = new BatchSourceContextImpl();
+    }
+  }
+
+  public boolean isUnbounded() {
+    return isUnbounded;
+  }
+
+  public Configuration getHadoopConf() {
+    if (hadoopConf == null) {
+      hadoopConf = new Configuration(false);
+    }
+    return hadoopConf;
+  }
+
+  public void prepareRun(PluginConfig config) {
+    // TODO: validate
+
+    for (Constructor<?> constructor : cdapPluginClass.getDeclaredConstructors()) {
+      constructor.setAccessible(true);
+      try {
+        cdapPluginObj = (T) constructor.newInstance(config);
+      } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+        LOG.error("Can not instantiate CDAP plugin class", e);
+        throw new IllegalStateException("Can not call prepareRun");
```

Comment:
Why do we create new exception instead of throwing existing one or wrapping it? (This applies to all error handling in this PR)

Reply:
Because I assumed we will only throw unchecked `IllegalStateException`s at least for now, until we will create our own `CustomException` and add it to the `throws` block.